src/DssDirectDepositHub.sol

```diff
@@ -0,0 +1,456 @@
+}
+
+contract DssDirectDepositHub {
+
+    // --- Auth ---
+    mapping (address => uint256) public wards;
+    function rely(address usr) external auth {
+        wards[usr] = 1;
+        emit Rely(usr);
+    }
+    function deny(address usr) external auth {
+        wards[usr] = 0;
+        emit Deny(usr);
+    }
+    modifier auth {
+        require(wards[msg.sender] == 1, "DssDirectDepositHub/not-authorized");
+        _;
+    }
+
+    enum Mode{ NORMAL, MODULE_CULLED, MCD_CAGED }
+    uint256             constant  RAY  = 10 ** 27;
+
+    VatLike      public immutable vat;
+    DaiJoinLike  public immutable daiJoin;
+    address      public           vow;
+    EndLike      public           end;
+
+    struct Ilk {
+        D3MPoolLike pool;   // Access external pool and holds balances
+        D3MPlanLike plan;   // How we calculate target debt
+        uint256     tau;    // Time until you can write off the debt [sec]
+        uint256     culled; // Debt write off triggered
+        uint256     tic;    // Timestamp when the pool is caged
+    }
+    mapping (bytes32 => Ilk) public ilks;
+    uint256                  public live = 1;
+
+    // --- Events ---
+    event Rely(address indexed usr);
+    event Deny(address indexed usr);
+    event File(bytes32 indexed ilk, bytes32 indexed what, address data);
+    event File(bytes32 indexed ilk, bytes32 indexed what, uint256 data);
+    event File(bytes32 indexed what, address data);
+    event Wind(bytes32 indexed ilk, uint256 amount);
+    event Unwind(bytes32 indexed ilk, uint256 amount);
+    event Reap(bytes32 indexed ilk, uint256 amt);
+    event Cage();
+    event Cage(bytes32 indexed ilk);
+    event Cull(bytes32 indexed ilk);
+    event Uncull(bytes32 indexed ilk);
+    event Quit(bytes32 indexed ilk, address indexed usr);
+    event Exit(bytes32 indexed ilk, address indexed usr, uint256 amt);
+
+    constructor(address vat_, address daiJoin_) public {
+        vat = VatLike(vat_);
+        daiJoin = DaiJoinLike(daiJoin_);
+        TokenLike(DaiJoinLike(daiJoin_).dai()).approve(daiJoin_, type(uint256).max);
+        VatLike(vat_).hope(daiJoin_);
+
+        wards[msg.sender] = 1;
+        emit Rely(msg.sender);
+    }
+
+    // --- Math ---
+    function _add(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x + y) >= x, "DssDirectDepositHub/overflow");
+    }
+    function _sub(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require((z = x - y) <= x, "DssDirectDepositHub/underflow");
+    }
+    function _mul(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        require(y == 0 || (z = x * y) / y == x, "DssDirectDepositHub/overflow");
+    }
+    function _min(uint256 x, uint256 y) internal pure returns (uint256 z) {
+        z = x <= y ? x : y;
+    }
+
+    // --- Administration ---
+    function file(bytes32 what, address data) external auth {
+        require(vat.live() == 1, "DssDirectDepositHub/no-file-during-shutdown");
```

Comment:
Do we need to have these live checks here? I don't think it provides any additional safety, and there could be some obscure edge case we are missing by disallowing calling this. I think we can remove both the `vat.live() == 1` and `live == 1` checks in all `file` functions.

Reply:
Yeah I'm not worried about the gas cost. To me some of these seem to prevent governance from being able to mess with the accounting indefinitely, but really governance has full control to do whatever it wants. In the case of governance misbehaving that should be handled by the ESM route and this hub de-authed anyways. I'm concerned about some legitimate edge case issue. We tend not to run `vat.live()` checks anywhere else although these seem to have been added sporadically in the previous D3M implementation.

@gbalabasquer do you remember why we have these `vat.live()` checks on the [`king` filing](https://github.com/makerdao/dss-direct-deposit/blob/master/src/DssDirectDepositAaveDai.sol#L220)? Does it make sense to keep them here when we don't do these checks to block `vow` / `end` filing anywhere else?

Reply:
It depends on if a change on a field can produce harm by malicious governance after the system is caged. Of course ESM can prevent it de-authorizing but in fact that ESM capability was designed to counteract the lack of `vat.live` or just `live` (in case End cages the module) checks for certain modules.
So a module that implements these checks directly would avoid the need of a keeper de-authorizing.